### PR TITLE
Fix writing pyramid

### PR DESCRIPTION
**Pull request by miriam-groeneveld**
Not every level was written with the correct data. Creating dzi files seems to have obfuscated the issue on GC.

### CODE REVIEW

File: panimg/models.py [hunk 1]
```diff
@@ -5,13 +5,24 @@
...
+    Tuple,
+)
 from uuid import UUID, uuid4
 
 import numpy as np
 import SimpleITK
-from pydantic import BaseModel, ConfigDict, field_validator
+from pydantic import BaseModel, ConfigDict
 from pydantic.dataclasses import dataclass
+from pydantic.functional_validators import field_validator
```
Comment (jmsmkn) [inline panimg/models.py:25]: Why the changes to this file? I don't see why the import needs to be moved or why the type annotations need to be added.

File: panimg/contrib/wsi_dcm_to_tiff/dcm_to_tiff.py [hunk 2]
```diff
@@ -228,36 +228,35 @@ def dcm_to_tiff(input_dir, output_path):
...
                 for col in range(cols):
                     im = np.array(getter(level, (col, row)))
                     yield im
 
-        for level in range(0, len(image.levels.levels)):
-            cols = int(math.ceil(image.levels[level].size.width / tile_size))
-            rows = int(math.ceil(image.levels[level].size.height / tile_size))
+        for level in image.levels:
+            cols = int(math.ceil(level.size.width / tile_size))
+            rows = int(math.ceil(level.size.height / tile_size))
```
Comment (jmsmkn) [inline panimg/contrib/wsi_dcm_to_tiff/dcm_to_tiff.py:241]: Could you add/update a test so that we're sure this does not regress? Are you certain that this fixes all issues with DICOM WSI?

File: panimg/models.py (hunk 1 above)
Comment (miriam-groeneveld) [reply panimg/models.py:25]: That was it, mypy was outdated. Updated it and reverted this change.

File: panimg/image_builders/tiff.py [hunk 3]
```diff
@@ -18,15 +18,18 @@
...
 
 try:
-    import openslide
+    if hasattr(os, "add_dll_directory"):
+        # Windows
+        OPENSLIDE_PATH = r"C:\Program Files\OpenSlide\bin"
+        with os.add_dll_directory(OPENSLIDE_PATH):
+            import openslide
+    else:
+        import openslide
```
Comment (pkcakeout) [inline panimg/image_builders/tiff.py:32]: Wow, this looks pretty random... why is this needed, or better said: Who needs this? Normally you should just setup your runtime environment correctly. This looks like very wild, random path guessing to me...

Comment (pkcakeout) [reply panimg/image_builders/tiff.py:32]: For the record, with "Setup your runtime" I mean:

~~~~~.bat
set PATH=%PATH%;C:\Program Files\OpenSlide\bin
python3 do-what-I-want.py
~~~~~

Comment (miriam-groeneveld) [reply panimg/image_builders/tiff.py:32]: that wasn't supposed to be committed...

Comment (pkcakeout) [reply panimg/image_builders/tiff.py:32]: :rofl: - ok, well spotted by me then... neverthelss, I _can_ see the utility. Maybe we want to add it to the docs somehow? We could _also_ provide a snippet like this to run as part of:

https://docs.python.org/3/using/cmdline.html?highlight=pythonpath#envvar-PYTHONSTARTUP

just some suggestions!

**Comment by miriam-groeneveld** [comment]
OpenSlide now supports DICOM, which I think is a much better solution than writng our own converter. These changes are not part of libvips yet. We can either:

- build libdicom, openslide and libvips from git master, as mentioned here: https://github.com/libvips/libvips/issues/1225. I've asked for an update on the expected release date.
- use openslide directly for the conversion
- wait for libvips (expectd to be added end of September)

I'll add an update once John Cupitt has replied, and just add the OpenSlide conversion if no answer by Monday.

**Comment by pkcakeout** [comment]
I honestly think that building `libvips` ourselves might be a better solution, especially because it seems clear that they will ultimately add the functionality. What I would do is to add the build-code right here in this repository so that rse-panimg still builds from sources. I can help with that if you want.

The disadvantage of adding `openslide` is that we add another hard dependency that we need to maintain ourselves. If we only rely on `libvips` our dependency graph stays leaner, with a short-term pain of having to maintain the build system.

**Comment by pkcakeout** [comment]
From a functional/implementation side: adding `openslide` directly is not a bad solution either of course!

**Comment by miriam-groeneveld** [comment]
This PR became a mess, I opened a new one with at least a fix for the issue: https://github.com/DIAGNijmegen/rse-panimg/pull/104
